internal/data/metric.go

```diff
@@ -0,0 +1,629 @@
+func NewHistogramBucketSlice(len int) []HistogramBucket {
+	origs := make([]otlpmetric.HistogramValue_Bucket, len)
+	wrappers := make([]HistogramBucket, len)
+	for i := range origs {
+		wrappers[i].orig = &origs[i]
+	}
+	return wrappers
+}
+
+func (hb *HistogramBucket) Count() uint64 {
+	// TODO: Remove conversion when proto and the generated code is fixed.
+	return uint64(hb.orig.Count)
+}
+
+func (hb *HistogramBucket) SetCount(v uint64) {
+	// TODO: Remove conversion when proto and the generated code is fixed.
+	hb.orig.Count = int64(v)
+}
+
+func (hb *HistogramBucket) Exemplar() HistogramBucketExemplar {
+	return hb.exemplar
+}
+
+func (hb *HistogramBucket) SetExemplar(v HistogramBucketExemplar) {
+	hb.exemplar = v
+}
+
+// HistogramBucket are example points that may be used to annotate aggregated Histogram values.
+// They are metadata that gives information about a particular value added to a Histogram bucket.
+//
+// Must use NewHistogramBucketExemplar* functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type HistogramBucketExemplar struct {
+	// Wrap OTLP HistogramValue_Bucket_Exemplar.
+	orig *otlpmetric.HistogramValue_Bucket_Exemplar
+
+	// Override a few fields. These fields are the source of truth. Their counterparts
+	// stored in corresponding fields of "orig" are ignored.
+	attachments Labels
+}
+
+func NewHistogramBucketExemplar() *HistogramBucketExemplar {
+	return &HistogramBucketExemplar{orig: &otlpmetric.HistogramValue_Bucket_Exemplar{}}
+}
+
+func (hbe *HistogramBucketExemplar) Value() float64 {
+	return hbe.orig.Value
+}
+
+func (hbe *HistogramBucketExemplar) SetValue(v float64) {
+	hbe.orig.Value = v
+}
+
+func (hbe *HistogramBucketExemplar) Timestamp() TimestampUnixNano {
+	return TimestampUnixNano(hbe.orig.TimestampUnixnano)
+}
+
+func (hbe *HistogramBucketExemplar) SetTimestamp(v TimestampUnixNano) {
+	hbe.orig.TimestampUnixnano = uint64(v)
+}
+
+func (hbe *HistogramBucketExemplar) Attachments() Labels {
+	return hbe.attachments
+}
+
+func (hbe *HistogramBucketExemplar) SetAttachments(v Labels) {
+	hbe.attachments = v
+}
+
+// SummaryDataPoint is a single data point in a timeseries that describes the time-varying
+// values of a Summary metric.
+//
+// Must use NewSummaryDataPoint* functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type SummaryDataPoint struct {
+	// Wrap OTLP SummaryDataPoint.
+	orig *otlpmetric.SummaryDataPoint
+
+	// Override a few fields. These fields are the source of truth. Their counterparts
+	// stored in corresponding fields of "orig" are ignored.
+	labels             Labels
+	valueAtPercentiles []*SummaryValueAtPercentile
+}
+
+func NewSummaryDataPoint() *SummaryDataPoint {
+	return &SummaryDataPoint{orig: &otlpmetric.SummaryDataPoint{}}
+}
+
+// NewSummaryDataPointSlice creates a slice of SummaryDataPoint that are correctly initialized.
+func NewSummaryDataPointSlice(len int) []SummaryDataPoint {
```

Comment:
For efficiency purposes this likely better to be a slice to pointers since SummaryDataPoint is not a very small struct (5 words):
```suggestion
func NewSummaryDataPointSlice(len int) []*SummaryDataPoint {
```

Reply:
The returned value (slice) is only 2 words, where do you see better efficiency, probably I miss some context here?

Reply:
Nit: slice is 3 words, but that was not what I was referring to. The slice itself is bigger because each element is the full struct, which is 5 words, instead of just pointer which is one word. That would make resizing and otherwise working with the slice slower. Arguably we probably are not going to do that in our codebase, so perhaps it doesn't really matter.

I may be wrong though, it is best to benchmark this, but until we have real code that uses this data structures it is not possible to write a benchmark that reflects real-world usage.

You can keep for now, but we may need to refactor it.